src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -415,10 +438,43 @@ internal SynthesizedAttributeData SynthesizeDebuggerStepThroughAttribute()
             return TrySynthesizeAttribute(WellKnownMember.System_Diagnostics_DebuggerStepThroughAttribute__ctor);
         }
 
-        internal SynthesizedAttributeData SynthesizeReadOnlyAttribute()
+        internal void EnsureIsReadOnlyAttributeExists(DiagnosticBag diagnostics, Location location, bool modifyCompilation)
         {
-            // PROTOTYPE(readonlyRefs) it is optional now as it will be generated in the next PR
-            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_ReadOnlyAttribute__ctor, isOptionalUse: true);
+            var isNeeded = CheckIfIsReadOnlyAttributeNeeded(diagnostics, location);
+
+            if (isNeeded && modifyCompilation)
+            {
+                NeedsGeneratedIsReadOnlyAttribute = true;
+            }
+        }
+
+        internal bool CheckIfIsReadOnlyAttributeNeeded(DiagnosticBag diagnosticsOpt, Location locationOpt)
+        {
+            var userDefinedAttribute = GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_IsReadOnlyAttribute);
+
+            if (userDefinedAttribute is MissingMetadataTypeSymbol)
+            {
+                if (Options.OutputKind == OutputKind.NetModule)
+                {
+                    if (diagnosticsOpt != null)
+                    {
+                        var errorReported = Binder.ReportUseSiteDiagnostics(userDefinedAttribute, diagnosticsOpt, locationOpt);
+                        Debug.Assert(errorReported);
+                    }
+                }
+                else
+                {
+                    return true;
+                }
+            }
+            else if (diagnosticsOpt != null && !userDefinedAttribute.InstanceConstructors.Any(constructor => constructor.ParameterCount == 0))
```

Comment:
This code, for example, doesn't check accessibility.

---
In reply to: [115102039](https://github.com/dotnet/roslyn/pull/18715#discussion_r115102039) [](ancestors = 115102039)